Enh/take along axis

bzah opened the pull request:
This PR adds `take_along_axis` that works similarly to [numpy's take_along_axis](https://numpy.org/doc/stable/reference/generated/numpy.take_along_axis.html).

- [x] Closes #3663 
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

# Credit
@zklaus for providing a working solution in [climix](https://git.smhi.se/climix/climix/-/blob/main/climix/dask_take_along_axis.py).

## Example of use
```python
from dask.array import take_along_axis

data: dask.array.Array 
top10_indices = data.argtopk(k=10, axis=-1)
top10 = take_along_axis(data, top10_indices, axis=-1)

# Equivalent to 
top10 = data.topk(k=10, axis=-1)
```

## Performances
This is just a basic benchmark I ran on my laptop to compare how this implementation prerforms against numpy's.
```python

import numpy as np
from dask.array import from_array
from dask.array.slicing import take_along_axis


random_arr = np.random.rand(1000,1000,1000)
dask_random_arr = from_array(random_arr)
top50 = dask_random_arr.argtopk(k=50, axis=0)
top50_np = top50.compute()

# Compute both argtopk and take_along_axis
%timeit take_along_axis(dask_random_arr, top50, axis=0).compute()
# 6.74 s ± 9.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit np.take_along_axis(random_arr, top50_np, axis=0)
# 864 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit take_along_axis(dask_random_arr, from_array(top50_np), axis=0).compute()
# 2.3 s ± 9.04 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


```

GPUtester commented:
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

bzah commented:
@zklaus, there are a few deviation from your original code:
- I removed sparse arrays as it's not a dependency for dask.
- I remove the [check of shapes](https://git.smhi.se/climix/climix/-/blob/main/climix/dask_take_along_axis.py#L35-L38) because it was breaking a use case. In particular, with this `assert` it break the following test *dask/array/tests/test_slicing.py::test_take_along_axis__indexing_twice_same_1darray*
- I removed the `meta` argument passed in `blockwise` as it was a sparse array. I suspect it some sort of optimization to give blockwise what is the expected shape but it's not really documented.
- Also some variable renaming and added types, docs and integration tests.

bzah commented:
Is there someone available to review/comment on this PR ?

zklaus commented:
Sorry that I don't have the bandwidth to interact properly :/ One thing that concerns me a little is that the test/benchmark isn't really testing much dask because it is conducted with a dask array that consists of a single numpy array as it's only chunk. In that case, performance would probably be expected to be slightly worse than numpy, but it's also not really representative.

Particularly, the point of the sparse array, iirc, was that there is one array for every chunk that covers *all* the chunks. If this is sparse, that's ok (or at least I thought so when writing it), but it is prohibitive if dense, so I don't think replacing the use of sparse with dense matrices is the way to go. If we want to avoid the use of sparse as a dependency, the aggregation step needs to be re-thought.

As a practical step forward, @bzah, could you do a larger test? Why not apply this to some larger climate data in at least a local cluster with several workers? I suspect that perhaps more than compute time, memory requirements without the use of sparse would balloon.

mrocklin commented:
@quasiben maybe this is something your team could review?

kgryte commented:
Related to this PR is a proposal to add `take_along_axis` to the Array API Standard: https://github.com/data-apis/array-api/issues/808.

If there are concerns regarding adding `take_along_axis` to Dask, it would be great to hear those concerns so that we can take those into account when formalizing expected behavior.
